### PR TITLE
[InstCombine] Fold `switch(zext/sext(X))` into `switch(X)`

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch folds `switch(zext/sext(X))` into `switch(X)`.
The original motivation of this patch is to optimize a pattern found in cvc5. For example:
```
  %bf.load.i = load i16, ptr %d_kind.i, align 8
  %bf.clear.i = and i16 %bf.load.i, 1023
  %bf.cast.i = zext nneg i16 %bf.clear.i to i32
  switch i32 %bf.cast.i, label %if.else [
    i32 335, label %if.then
    i32 303, label %if.then
  ]

if.then:                                          ; preds = %entry, %entry
  %d_children.i.i = getelementptr inbounds %"class.cvc5::internal::expr::NodeValue", ptr %0, i64 0, i32 3
  %cmp.i.i.i.i.i = icmp eq i16 %bf.clear.i, 1023
  %cond.i.i.i.i.i = select i1 %cmp.i.i.i.i.i, i32 -1, i32 %bf.cast.i
```
`%cmp.i.i.i.i.i` always evaluates to false because `%bf.clear.i` can only be 335 or 303.
Folding `switch i32 %bf.cast.i` to `switch i16 %bf.clear.i` will help `CVP` to handle this case.
See also https://github.com/llvm/llvm-project/pull/76928#issuecomment-1877055722.

Compile-time impact: http://llvm-compile-time-tracker.com/compare.php?from=7954c57124b495fbdc73674d71f2e366e4afe522&to=502b13ed34e561d995ae1f724cf06d20008bd86f&stat=instructions:u
|stage1-O3|stage1-ReleaseThinLTO|stage1-ReleaseLTO-g|stage1-O0-g|stage2-O3|stage2-O0-g|stage2-clang|
|--|--|--|--|--|--|--|
|+0.03%|+0.06%|+0.07%|+0.00%|-0.02%|-0.03%|+0.02%|

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
Why is this in SimplifyCFG? This seems like an extension of the existing switch width fold in InstCombine: https://github.com/llvm/llvm-project/blob/8b2bdfbca7c1db272e4e703445f5626b4bc4b9d3/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L3224

**Comment by dtcxzyw** [comment]
> Why is this in SimplifyCFG? This seems like an extension of the existing switch width fold in InstCombine:
> 
> https://github.com/llvm/llvm-project/blob/8b2bdfbca7c1db272e4e703445f5626b4bc4b9d3/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L3224

IIRC, `InstCombine` cannot change the dominator tree.

See the following case:
```
 define i1 @test_switch_with_sext_unreachable_case(i16 %a, i1 %b, i1 %c) {
 entry:
   %a.ext = sext i16 %a to i32
   switch i32 %a.ext, label %sw.default [
   i32 37, label %sw.bb
   i32 38, label %sw.bb
   i32 39, label %sw.bb
   i32 -65499, label %sw.bb2
   ]

 sw.bb:
   ret i1 %b
 sw.bb2:
   ret i1 false
 sw.default:
   ret i1 %c
 }
```
In this case, the edge `i32 -65499, label %sw.bb2` should be eliminated.

**Comment by nikic** [comment]
@dtcxzyw But won't that edge get eliminated by SimplifyCFG anyway?

**Comment by dtcxzyw** [comment]
> @dtcxzyw But won't that edge get eliminated by SimplifyCFG anyway?

Okay, I have got your point.

**Comment by dtcxzyw** [comment]
> Why is this in SimplifyCFG? This seems like an extension of the existing switch width fold in InstCombine:
> 
> https://github.com/llvm/llvm-project/blob/8b2bdfbca7c1db272e4e703445f5626b4bc4b9d3/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L3224

Done.

**Comment by dtcxzyw** [comment]
> <!DOCTYPE html><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; caret-color: rgb(31, 35, 40); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Compile-time impact:<span class="Apple-converted-space"> </span><a href="http://llvm-compile-time-tracker.com/compare.php?from=7954c57124b495fbdc73674d71f2e366e4afe522&amp;to=502b13ed34e561d995ae1f724cf06d20008bd86f&amp;stat=instructions:u" rel="nofollow" style="box-sizing: border-box; background-color: transparent; color: var(--fgColor-accent, var(--color-accent-fg)); text-decoration: underline; text-underline-offset: 0.2rem;">http://llvm-compile-time-tracker.com/compare.php?from=7954c57124b495fbdc73674d71f2e366e4afe522&amp;to=502b13ed34e561d995ae1f724cf06d20008bd86f&amp;stat=instructions:u</a></p>
stage1-O3 | stage1-ReleaseThinLTO | stage1-ReleaseLTO-g | stage1-O0-g | stage2-O3 | stage2-O0-g | stage2-clang
-- | -- | -- | -- | -- | -- | --
+0.03% | +0.06% | +0.07% | +0.00% | -0.02% | -0.03% | +0.02%

Does `SimplifyCFG` revert this transform?

**Comment by nikic** [comment]
@dtcxzyw The compile-time increases are probably due to increased code size: http://llvm-compile-time-tracker.com/compare.php?from=7954c57124b495fbdc73674d71f2e366e4afe522&to=502b13ed34e561d995ae1f724cf06d20008bd86f&stat=size-text As long as the code size increase itself is not a regression, this is fine.

File: llvm/test/Transforms/InstCombine/switch-zext-sext.ll [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+; CHECK-NEXT:      i16 38, label [[SW_BB]]
+; CHECK-NEXT:      i16 39, label [[SW_BB]]
+; CHECK-NEXT:    ]
+; CHECK:       sw.bb:
+; CHECK-NEXT:    ret i1 [[B]]
+; CHECK:       sw.default:
+; CHECK-NEXT:    ret i1 [[C]]
+;
+entry:
+  %a.ext = sext i16 %a to i32
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/switch-zext-sext.ll:47]: I'd like to see a test where the sign extension matters.

**Review by nikic — APPROVED** [review]
LGTM if you're happy with the diffs :)

File: llvm/test/Transforms/InstCombine/switch-zext-sext.ll (hunk 1 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/switch-zext-sext.ll:47]: Done.

**Pull request merged**